### ISSUE TITLE
Mapped types does not work with generic type arguments extending an object

### ISSUE THREAD

**Issue by HosseinAgha** [microsoft/TypeScript#35647]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:** mapped types generics bug

**Code**

I can confirm that this does __not__ work:  
```typescript
  const genericFunc = async <Schema extends { _id: string }>() => {
    type a = { [K in keyof Schema]: Schema[K] };
    const b: a = { _id: 'sd' };
  }
```
And this does work:  
```typescript
  type Schema2 = { _id: string };
  type a2 = { [K in keyof Schema2]: Schema2[K] };
  const b2: a2 = { _id: 'sd' };
```

**Expected behavior:**
IMO TypeScript should assume generic parameter above is something like this:  
```typescript
interface Schema {
  _id: string;
  [K: string]: any;
}
```
and does not throw errors.

**Playground Link:** [example link](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=11&pc=2#code/C4TwDgpgBAygxgCwgWwIZQLxQN5QPoCWAJgFxQDOwATgQHYDmUAvgNwBQbA9J1AOoD2VANbkSbUJCgAFVFWAFUAG3hI0ARkzTZ8pQB4VKVAD52cfrUpQARgCYyMuQuWJDGrLkKkoAcnJFvzOxcPAAi-BDktN7AUADugkJQABRWAK4xBDHkCPypikQAlGIAZqm0cPLmUPQQtBA0cACyEMA5RLyZCACiAB7AtUTkugAqUBB9A+SwLmhGSQU4bFDLUBLQDjrOqqg2mhtOIyZLK2YWMVZq9tpOBmi77vjEZL7+gWxMQA)
**Related Issues:** I'm one of the maintainers of the `@types/mongodb` in definitely typed. This is the original issue https://github.com/DefinitelyTyped/DefinitelyTyped/issues/39358

**Comment by AnyhowStep** [comment]
https://github.com/microsoft/TypeScript/issues/35077

**Comment by RyanCavanaugh** [comment]
This is a correct error, because the code makes a promise it's not keeping - specifically that for any `T`, a valid `Partial<T>` is produced.

```ts
// Doesn't work (but it should):
function genericMethodWithExtends<T extends Schema>(): Partial<T> {
    type PartialSchema2 = Partial<T>;
    const b1: PartialSchema2 = { _id: 'sd' };
    return b1;
}

type MySchema = { id?: "foo" }
const k = genericMethodWithExtends<MySchema>();
// The only legal inhabitants of 'v' are "foo" and "undefined", but
// v has value "sd" instead.
const v = k.id;
```

https://twitter.com/SeaRyanC/status/1121995986862084096

**Comment by fatcerberus** [comment]
> Every drink in a bar comes in a glass (the constraint), but the waiter (the function) can't just bring you an empty glass (the return value) if you asked for a beer (T)

Well I mean... maybe the barkeep is just cutting you off because you've already ordered too many drinks that night?

**Comment by HosseinAgha** [comment]
@RyanCavanaugh I understand the classic OOP argument:  

```ts
class Cat extends Animal {}

// this is correct as Cat is bigger than Animal so we don't have a wrong property access
const a: Animal = new Cat();
// but this is wrong as you may access a property on Cat that does not exist on Animal
const c: Cat = new Animal();
```

You are right, I did not consider [`string/numeric literal types`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types) when I made my argument. In other words I considered `string` (and any other Basic Type) not extendable.  

So in a language without `string/numeric literal types` the `T extends string` means `T == string`.

__I still don't understand why the following throws:__ 
```ts
function genericMethodWithExtends<T extends 'sd'>(): T {
    return 'sd';
}
```

We've reached the end of type tree here, nobody can extend 'sd'.

**Comment by AnyhowStep** [comment]
```ts
function genericMethodWithExtends<T extends 'sd'>(): T {
    return 'sd';
}
```

There's `never`, `"sd" & { aBrandType : void }`, etc.

**Comment by HosseinAgha** [comment]
Thank you @AnyhowStep.  
1) looks right. Although `"sd" & { aBrandType : void }` is a little bit strange (I guess everything is an object in JavaScript 😃)
---
2) what does `const a = { p1: 12, p2: 'sdsdf' } as const` mean?
---
3) `T could be { _id: string }`: isn't this enough to resolve this conditional type `keyof T extends '_id' ? string : number;`?

**Comment by AnyhowStep** [comment]
> what does `const a = { p1: 12, p2: 'sdsdf' } as const` mean?

It's a "const assertion"

http://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions

-----

> `T could be { _id: string }`: isn't this enough to resolve this conditional type `keyof T extends '_id' ? string : number;`?

If `T` is `{ _id : string, x : number }`, then `keyof T` is `"_id"|"x"`.

`"_id"|"x"` does not extend `"_id"`.

It is possible for the conditional type to evaluate to the `true` branch. Also possible to evaluate to the `false` branch. It depends on what `T` is.

**Comment by jimbuck** [comment]
I totally follow the original question and answer (all Beers are Beverages, all Beverages are not Beers), but not @RyanCavanaugh's answer (why is there `id` **and** `_id`). How is `Partial<T>` supposed to fit in with all of this? Given the following code:

[Playground][playground]

```ts
interface Schema {
  id: string;
}

function genericMethodWithExtends<T extends Schema>(): void {
  type PartialOfT      = Partial<T>;
  type PartialOfSchema = Partial<Schema>;
  
  const b1: PartialOfSchema = { id: 'sd' }; // <- This is valid.
  const b2: PartialOfT      = { id: 'sd' }; // <- This is not. Why?
}

type MySchema = { name: string, id: string }
genericMethodWithExtends<MySchema>();
```

Should the assignment to a `Partial<T>` fail? If `Partial<T>` is ["a type that represents all subsets of a given type"][partialtdocs] then surely the assignment to `b1` (above) should be valid. Or are generic subclasses not supported by `Partial<T>` currently? Or is there a better way to model these types?

**To continue the glass/beer analogy**: _The waiter gave me my beer in a glass and I promise to return the glass and/or the beer._

And for reference, the types created above show up in intellisense like so:
```ts
type PartialOfT = { [P in keyof T]?: T[P] | undefined; }
type PartialOfSchema = { id?: string | undefined; }
```

[playground]: http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=14&pc=38#code/JYOwLgpgTgZghgYwgAgMoIBYQLZ2QbwFgAoZZYAEwC5kBnMKUAcwG4SBfEkmAVxATDAA9iGRMIIaMAQBZCGAxCKAdWAKAogA9IICrQA8AFWQRtEvWkw44APgAUAShoA3IZQIkyYAJ4AHFAAKcFCCcAA2APIwxmSxALzIQSHA4UY2bKTIPv6JwaGRMOhYuMgJSfn6Rdbpnsi1CCL0yABGAIw05SkFVSUJ+OTUyADktBRDyOwsyAD008j6ALTIhhjAtOTrzuGUAHT1jWAtAEwdeV1RMbGlBAM0I2MTU7PzSytrG8ggQmA7yMoY3gA-BwuMRsigZN4eng+p84NgIDR6IwQEwADS3OgMZgTEjiSSMWTyRQqNQYLQ6PT6SHQ+wOFhAA
[partialtdocs]: https://www.typescriptlang.org/docs/handbook/utility-types.html#partialt

**Comment by AnyhowStep** [comment]
Consider `T = { id : "lol" }`.

```ts
type PartialOfT = { id? : "lol"|undefined };
const b2: PartialOfT = { id : "sd" }; //Makes sense this is an error
```

**Comment by HosseinAgha** [comment]
@jimbuck as @AnyhowStep said the problem with your code again is that you should also consider that `string` can have infinite sub types. so `"sd"` is a `string` and  "lol"` is a `string` but `"sd" != "lol"`.

**Comment by jimbuck** [comment]
Ahh, the thing that got me was the fact that when you extend an interface/type you can "override" a `string` property with a string literal. I've updated my code to highlight this a little bit better:

[Playground][playground]

```ts
interface Schema {
  id: string;
}

function genericMethodWithExtends<T extends Schema>(): void {
  type PartialOfT      = Partial<T>;
  type PartialOfSchema = Partial<Schema>;
  
  const b1: PartialOfSchema = { id: 'sd' }; // <- This is valid.
  const b2: PartialOfT      = { id: 'sd' }; // <- For T this is supposed to be 'lol', hence the error.
}

interface MySchema extends Schema {
  id: 'lol' // This can be "overriden" since 'lol' is a string (no rules broken)
}
genericMethodWithExtends<MySchema>();
```

I know that `'lol'` is still a `string`, but it really is changing the type of the property on the interface, which is a little annoying to me. Either way, thank you for providing some insight on this scenario!!

[playground]: http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=16&pc=38#code/JYOwLgpgTgZghgYwgAgMoIBYQLZ2QbwFgAoZZYAEwC5kBnMKUAcwG4SBfEkmAVxATDAA9iGRMIIaMAQBZCGAxCKAdWAKAogA9IICrQA8AFWQRtEvWkw44APgAUAShoA3IZQIkyYAJ4AHFAAKcFCCcAA2APIwxmSxALzIQSHA4UY2bKTIPv6JwaGRMOhYuMgJSfn6Rdbpnsi1CCL0yABGAIw05SkFVSUJ+OTUyADktBRDyOwsyAD008j6ALTIhhjAtOTrzuGUAHT1jWAtAEwdeV1RMbGlBAM0I2MTU7PzSwBiQlDLWavra3Q8vl8QloEAoWSELRQQzCQjCQwANMgsPwUAoUNAoB89sROMQSKBILBECgZN4enhTDoLOSPJlKHcYXCZnMVn8EHBRM0UAAiITODGUCTcuigJDDRnjP54eiMEBMZB2EAQqA8MIQdbNTEAawkDg4JHEkkYsnkihUagwWipBlJ5PsDhYQA

**Comment by AnyhowStep** [comment]
It's not even about interface subtyping.

type `{ id : "lol" }` is a subtype of `Schema`.

Even though `{ id : "lol" }` is not an `interface` with an explicit `extends Schema` annotation.
